### PR TITLE
Check Privacy policy or T&C

### PR DESCRIPTION
**Pull request by Poppe-A**
#8282

### CODE REVIEW

File: libs/model/src/lib/user.ts [hunk 1]
```diff
@@ -45,6 +45,13 @@ export interface PublicUser {
   lastName?: string;
   orgId?: string;
   hideEmail: boolean;
+  legalTerms?: {
+    privacyPolicy: boolean;
+    tc: {
+      festival?: boolean;
+      catalog?: boolean;
```
[the remark this answers is not in the document]
Comment (cams-goussale) [reply libs/model/src/lib/user.ts]: Absolutely yes, we want to be able to update T&C and have users accept the new version again

File: libs/user/src/lib/auth/pages/checkPolicyAndTerms/checkPolicyAndTerms.component.html [hunk 2]
```diff
@@ -0,0 +1,26 @@
+<a app-logo routerLink="/"></a>
+
+<h1>Almost there!</h1>
+<section class="surface">
```
Comment (phpgeekfr) [inline libs/user/src/lib/auth/pages/checkPolicyAndTerms/checkPolicyAndTerms.component.html:4]: section should contain an heading https://www.notion.so/cascade8/Code-conventions-0dc3a5db4f1045f68f1587f9a88bcce2

**Review by phpgeekfr — COMMENTED** [review]
update `libs/devops/src/lib/firebase-utils/anonymize/lib.ts`

using same logic as : 

```
const privacyPolicy = { date: new Date(), ip: fakeIp() };
```

File: firestore.rules [hunk 3]
```diff
@@ -79,8 +79,6 @@ service cloud.firestore {
       	&& userId() == incomingData().uid
         && isNotUpdatingId('uid', existingData().uid)
         && isNotUpdatingField(['email'])
-        && (isNotUpdatingField(['privacyPolicy', 'date']) || isCreatingField(['privacyPolicy', 'date']))
-        && (isNotUpdatingField(['privacyPolicy', 'ip']) || isCreatingField(['privacyPolicy', 'ip']))
```
Comment (phpgeekfr) [inline firestore.rules:83]: You think that there is no point in updating this rule and prefer to remove it completely ?

Comment (phpgeekfr) [reply firestore.rules:83]: lets discuss it tomorrow

File: libs/user/src/lib/auth/pages/checkPolicyAndTerms/checkPolicyAndTerms.component.ts [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+  async accept() {
+    const uid = this.service.uid;
+    const legalTerms = await this.service.getLegalTerms();
+    const privacyPolicy = legalTerms;
+    const termsAndConditions = {
+      [this.app]: legalTerms
+    };
+
+    await this.service.update({ uid, termsAndConditions, privacyPolicy });
+    location.replace('/')
```
Comment (phpgeekfr) [inline libs/user/src/lib/auth/pages/checkPolicyAndTerms/checkPolicyAndTerms.component.ts:43]: you should be able to do `this.router.navigate(['/c/o']);`

If you look at http://localhost:4200/c/o/account/profile/view/settings for example, when you update your name, `ProfileViewComponent` is updated with correct value using `this.user$ = this.authService.profile$;`

Comment (phpgeekfr) [reply libs/user/src/lib/auth/pages/checkPolicyAndTerms/checkPolicyAndTerms.component.ts:43]: lets check this together

File: libs/user/src/lib/auth/pages/checkPolicyAndTerms/checkPolicyAndTerms.component.ts [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+  async accept() {
+    const uid = this.service.uid;
+    const legalTerms = await this.service.getLegalTerms();
+    const privacyPolicy = legalTerms;
+    const termsAndConditions = {
+      [this.app]: legalTerms
+    };
+
+    await this.service.update({ uid, termsAndConditions, privacyPolicy });
```
Comment (phpgeekfr) [inline libs/user/src/lib/auth/pages/checkPolicyAndTerms/checkPolicyAndTerms.component.ts:42]: you don't need to provide `uid` here

File: apps/festival/festival-e2e/src/integration/authentification/login.spec.ts [hunk 6]
```diff
@@ -26,8 +27,8 @@ describe('Login tests', () => {
     maintenance.start();
     adminAuth.deleteAllTestUsers();
     firestore.clearTestData();
-    adminAuth.createUser({ uid: user.uid, email: user.email });
-    adminAuth.updateUser({ uid: user.uid, update: { emailVerified: true } });
+    adminAuth.createUser({ uid: user.uid, email: user.email, emailVerified: true });
+    firestore.update([{docPath: `users/${user.uid}`, field: 'termsAndConditions', value: {} }])
```
Comment (phpgeekfr) [inline apps/festival/festival-e2e/src/integration/authentification/login.spec.ts:31]: bad indent and missing `;`

and you are updating user before creating it

File: libs/testing/cypress/browser/src/lib/support.ts [hunk 7]
```diff
@@ -1,5 +1,4 @@
 import { OrgActivity, Territory, PublicUser, Module } from '@blockframes/model';
```
Comment (phpgeekfr) [inline libs/testing/cypress/browser/src/lib/support.ts:1]: browserAuth is used in this file

File: libs/user/src/lib/auth/pages/checkPolicyAndTerms/checkPolicyAndTerms.module.ts [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+import { MatCheckboxModule } from '@angular/material/checkbox';
+
+// Component
+import { CheckPolicyAndTermsComponent } from './checkPolicyAndTerms.component';
+
+@NgModule({
+  declarations: [CheckPolicyAndTermsComponent],
+  imports: [
+    CommonModule,
+    FlexLayoutModule,
```
Comment (phpgeekfr) [inline libs/user/src/lib/auth/pages/checkPolicyAndTerms/checkPolicyAndTerms.module.ts:22]: I think you don't need FlexLayoutModule

Comment (phpgeekfr) [reply libs/user/src/lib/auth/pages/checkPolicyAndTerms/checkPolicyAndTerms.module.ts:22]: maybe also other modules ?

File: libs/user/src/lib/auth/service.ts [hunk 9]
```diff
@@ -27,6 +27,13 @@ import { doc, DocumentReference, getDoc, writeBatch } from 'firebase/firestore';
 
 @Injectable({ providedIn: 'root' })
 export class AuthService extends BlockframesAuth<User> implements OnDestroy {
+  // If we update the privacy policy or the T&C, we need the update the publishing dates
+  readonly privacyPolicyDate = new Date('Thu Jun 23 2022 15:14:36 GMT+0200');
+  readonly termsAndConditionsDate = {
```
Comment (phpgeekfr) [inline libs/user/src/lib/auth/service.ts:32]: type it with `Partial<Record<App, Date>>`

File: libs/devops/src/lib/firebase-utils/anonymize/lib.ts [hunk 10]
```diff
@@ -72,8 +73,13 @@ function processUser<T extends User | PublicUser>(u: T): T {
   const firstName = faker.name.firstName();
   const lastName = faker.name.lastName();
   const email = fakeEmail(firstName);
-  const privacyPolicy = { date: new Date(), ip: fakeIp() };
-  return { ...u, firstName, lastName, email, privacyPolicy };
+  const legalTerms = { date: new Date(), ip: fakeIp() };
+  const privacyPolicy = legalTerms;
+  const termsAndConditions = {};
+  for (const appName of app) {
```
Comment (phpgeekfr) [inline libs/devops/src/lib/firebase-utils/anonymize/lib.ts:79]: use `getAllAppsExcept(['crm'])`

**Review by phpgeekfr — COMMENTED** [review]
please update test `describe('Test USERS anonymization function', () => {`  in `libs/devops/src/lib/firebase-utils/anonymize/lib.spec.ts`

File: libs/devops/src/lib/firebase-utils/anonymize/lib.spec.ts [hunk 11]
```diff
@@ -53,7 +53,22 @@ describe('Test USERS anonymization function', () => {
...
+            date: new Date()
+          },
+          catalog: {
+            ip: '127.0.0.1',
+            date: new Date()
+          },
+          financiers: {
+            ip: '127.0.0.1',
+            date: new Date()
+          },
```
Comment (phpgeekfr) [inline libs/devops/src/lib/firebase-utils/anonymize/lib.spec.ts:71]: also add 

```
expect(doc.content.termsAndConditions.festival.ip).not.toEqual(userRecord.content.termsAndConditions.festival.ip);
```

for the 3 apps

File: libs/user/src/lib/auth/pages/checkPolicyAndTerms/checkPolicyAndTerms.module.ts [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+import { CheckPolicyAndTermsComponent } from './checkPolicyAndTerms.component';
+
+@NgModule({
+  declarations: [CheckPolicyAndTermsComponent],
+  imports: [
+    CommonModule,
+    AppLogoModule,
+    ReactiveFormsModule,
+    AppPipeModule,
+    ImageModule,
```
[the remark this answers is not in the document]
Comment (Poppe-A) [reply libs/user/src/lib/auth/pages/checkPolicyAndTerms/checkPolicyAndTerms.module.ts:24]: ImageModule is needed for the background

File: libs/user/src/lib/auth/guard/legal-terms.guard.ts [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+        const termsAndConditionsDate = user.termsAndConditions?.[this.app]?.date;
+        if (
+          this.app !== 'crm' && (
+          !privacyPolicyDate || 
+          privacyPolicyDate < this.service.privacyPolicyDate ||
+          !termsAndConditionsDate ||
+          termsAndConditionsDate < this.service.termsAndConditionsDate[this.app])
+          ) {
+          return this.router.createUrlTree(['/auth/checkPrivacyAndTerms']);
+        }
```
Comment (phpgeekfr) [inline libs/user/src/lib/auth/guard/legal-terms.guard.ts:33]:
```suggestion
        if(this.app === 'crm')  return true;

        const latestPrivacyPolicy = this.service.privacyPolicyDate;
        const latestTermsAndConditions = this.service.termsAndConditionsDate[this.app];
        const userPrivacyPolicy = user.privacyPolicy?.date;
        const userTermsAndConditions = user.termsAndConditions?.[this.app]?.date;

        const invalidPrivacyPolicy = !userPrivacyPolicy ||  userPrivacyPolicy < latestPrivacyPolicy;
        const invalidTermsAndConditions =  !userTermsAndConditions || userTermsAndConditions <  latestTermsAndConditions;
        if ( invalidPrivacyPolicy  || invalidTermsAndConditions ) {
          return this.router.createUrlTree(['/auth/checkPrivacyAndTerms']);
        }
```

File: libs/user/src/lib/auth/guard/no-legal-terms.guard.ts [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+        const termsAndConditionsDate = user.termsAndConditions?.[this.app]?.date;
+        if (
+          this.app !== 'crm' && (
+          !privacyPolicyDate || 
+          privacyPolicyDate < this.service.privacyPolicyDate ||
+          !termsAndConditionsDate ||
+          termsAndConditionsDate < this.service.termsAndConditionsDate[this.app])
+          ) {
+          return true;
+        }
```
Comment (phpgeekfr) [inline libs/user/src/lib/auth/guard/no-legal-terms.guard.ts:33]: same logic as other guard

File: libs/user/src/lib/auth/pages/checkPolicyAndTerms/checkPolicyAndTerms.component.ts [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+    this.form.get('termsOfUse').setValue(isTCAccepted);
+    this.form.get('privacyPolicy').setValue(isPrivacyAccepted);
+  }
+
+  async accept() {
+    const legalTerms = await this.service.getLegalTerms();
+    const privacyPolicy = legalTerms;
+    const termsAndConditions = {
+      [this.app]: legalTerms
+    };
```
Comment (phpgeekfr) [inline libs/user/src/lib/auth/pages/checkPolicyAndTerms/checkPolicyAndTerms.component.ts:43]: you should retreive existing user `termsAndConditions` for other apps and merge it here

File: libs/user/src/lib/auth/pages/checkPolicyAndTerms/checkPolicyAndTerms.component.scss [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+  }
+}
+
+@include layout-bp(lt-md) {
+  .surface {
+    width: 70%;
+  }
+}
+
+@include layout-bp(lt-sm) {
```
Comment (phpgeekfr) [inline libs/user/src/lib/auth/pages/checkPolicyAndTerms/checkPolicyAndTerms.component.scss:56]: still a bit overlapping with  
image

image

File: libs/user/src/lib/auth/guard/no-legal-terms.guard.ts [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+    private router: Router,
+    private service: AuthService,
+    @Inject(APP) private app: App
+  ) { }
+
+  canActivate() {
+    return this.service.profile$.pipe(
+      switchMap(() => this.service.getValue()),
+      map(user  => {
+        if (this.app === 'crm') return this.router.createUrlTree(['/c/o']);;
```
Comment (phpgeekfr) [inline libs/user/src/lib/auth/guard/no-legal-terms.guard.ts:23]:
```suggestion
        if (this.app === 'crm') return this.router.createUrlTree(['/c/o']);
```

File: libs/user/src/lib/auth/guard/no-legal-terms.guard.ts [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+        const latestPrivacyPolicy = this.service.privacyPolicyDate;
+        const latestTermsAndConditions = this.service.termsAndConditionsDate[this.app];
+        const userPrivacyPolicy = user.privacyPolicy?.date;
+        const userTermsAndConditions = user.termsAndConditions?.[this.app]?.date;
+        const invalidPrivacyPolicy = !userPrivacyPolicy ||  userPrivacyPolicy < latestPrivacyPolicy;
+        const invalidTermsAndConditions =  !userTermsAndConditions || userTermsAndConditions <  latestTermsAndConditions;
+        if ( invalidPrivacyPolicy  || invalidTermsAndConditions ) {
+          return true;
+        }
```
Comment (phpgeekfr) [inline libs/user/src/lib/auth/guard/no-legal-terms.guard.ts:32]:
```suggestion
        if ( invalidPrivacyPolicy  || invalidTermsAndConditions )  return true;
```

File: libs/user/src/lib/auth/guard/no-legal-terms.guard.ts [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+    return this.service.profile$.pipe(
+      switchMap(() => this.service.getValue()),
+      map(user  => {
+        if (this.app === 'crm') return this.router.createUrlTree(['/c/o']);;
+        const latestPrivacyPolicy = this.service.privacyPolicyDate;
+        const latestTermsAndConditions = this.service.termsAndConditionsDate[this.app];
+        const userPrivacyPolicy = user.privacyPolicy?.date;
+        const userTermsAndConditions = user.termsAndConditions?.[this.app]?.date;
+        const invalidPrivacyPolicy = !userPrivacyPolicy ||  userPrivacyPolicy < latestPrivacyPolicy;
+        const invalidTermsAndConditions =  !userTermsAndConditions || userTermsAndConditions <  latestTermsAndConditions;
```
Comment (phpgeekfr) [inline libs/user/src/lib/auth/guard/no-legal-terms.guard.ts:29]: maybe a bit more spacing

File: libs/testing/cypress/browser/src/lib/helpers.ts [hunk 20]
```diff
@@ -1,4 +1,4 @@
-import { PublicUser, PermissionsDocument, Organization } from '@blockframes/model';
+import { PublicUser, PermissionsDocument, Organization, User } from '@blockframes/model';
```
Comment (phpgeekfr) [inline libs/testing/cypress/browser/src/lib/helpers.ts:1]: PublicUser not used

File: libs/user/src/lib/auth/guard/no-legal-terms.guard.ts [hunk 21]
```diff
@@ -0,0 +1,37 @@
...
+        if ( invalidPrivacyPolicy  || invalidTermsAndConditions ) {
+          return true;
+        }
+        return this.router.createUrlTree(['/c/o']);;
```
Comment (phpgeekfr) [inline libs/user/src/lib/auth/guard/no-legal-terms.guard.ts:33]:
```suggestion
        return this.router.createUrlTree(['/c/o']);
```

File: libs/user/src/lib/auth/guard/legal-terms.guard.ts [hunk 22]
```diff
@@ -0,0 +1,39 @@
...
+        if (this.app === 'crm') return true;
+        const latestPrivacyPolicy = this.service.privacyPolicyDate;
+        const latestTermsAndConditions = this.service.termsAndConditionsDate[this.app];
+        const userPrivacyPolicy = user.privacyPolicy?.date;
+        const userTermsAndConditions = user.termsAndConditions?.[this.app]?.date;
+
+        const invalidPrivacyPolicy = !userPrivacyPolicy ||  userPrivacyPolicy < latestPrivacyPolicy;
+        const invalidTermsAndConditions =  !userTermsAndConditions || userTermsAndConditions <  latestTermsAndConditions;
+        
+        if ( invalidPrivacyPolicy  || invalidTermsAndConditions ) {
```
Comment (phpgeekfr) [inline libs/user/src/lib/auth/guard/legal-terms.guard.ts:32]: bad formatting

File: libs/user/src/lib/auth/guard/no-legal-terms.guard.ts [hunk 23]
```diff
@@ -0,0 +1,38 @@
...
+
+        const latestPrivacyPolicy = this.service.privacyPolicyDate;
+        const latestTermsAndConditions = this.service.termsAndConditionsDate[this.app];
+        const userPrivacyPolicy = user.privacyPolicy?.date;
+        const userTermsAndConditions = user.termsAndConditions?.[this.app]?.date;
+
+        const invalidPrivacyPolicy = !userPrivacyPolicy ||  userPrivacyPolicy < latestPrivacyPolicy;
+        const invalidTermsAndConditions =  !userTermsAndConditions || userTermsAndConditions <  latestTermsAndConditions;
+        
+        if ( invalidPrivacyPolicy  || invalidTermsAndConditions ) return true;
```
Comment (phpgeekfr) [inline libs/user/src/lib/auth/guard/no-legal-terms.guard.ts:33]: bad formatting

**Review by phpgeekfr — APPROVED** [review]

**Pull request merged**